### PR TITLE
Add in landsat search and submit functionality

### PR DESCRIPTION
**Pull request by jhkennedy**
Scene not in tile list; raises `AssertionError` 
```
python landsat/src/main.py LC09_L1TP_077023_20240124_20240124_02_T1
```
scene in tile list; should find 65 pairs
```
python landsat/src/main.py LC08_L1TP_138041_20240128_20240207_02_T1
```

### CODE REVIEW

File: landsat/src/main.py [hunk 1]
```diff
@@ -1,11 +1,202 @@
...
+ITS_LIVE_BUCKET = 'its-live-data'
+ITS_LIVE_TILES_KEY = 'catalog_geojson/landsatOLI/v2_LandsatOLI_tiles.json'
+
+
+STAC_CLIENT = pystac_client.Client
+LANDSAT_STAC_API = 'https://landsatlook.usgs.gov/stac-server'
+LANDSAT_CATALOG = pystac_client.Client.open(LANDSAT_STAC_API)
+LANDSAT_COLLECTION = 'landsat-c2l1'
+
+MAX_PAIR_SEPERATION = 544  # days
```
Comment (jtherrmann) [inline landsat/src/main.py:28]:
```suggestion
MAX_PAIR_SEPARATION = 544  # days
```

File: landsat/src/main.py [hunk 2]
```diff
@@ -1,11 +1,202 @@
...
+    collection = LANDSAT_CATALOG.get_collection(LANDSAT_COLLECTION)
+    item = collection.get_item(scene)
+    # TODO: raise specific errors instead of asserts
+    assert item.properties['eo:cloud_cover'] < max_cloud_cover
+    assert item.properties['view:off_nadir'] == 0
+
+
+def get_landsat_pairs_for_reference_scene(
+    reference: str,
+    max_pair_separation: timedelta = timedelta(days=MAX_PAIR_SEPERATION),
```
Comment (jtherrmann) [inline landsat/src/main.py:75]:
```suggestion
    max_pair_separation: timedelta = timedelta(days=MAX_PAIR_SEPARATION),
```

File: landsat/src/main.py [hunk 3]
```diff
@@ -1,11 +1,202 @@
...
+    jobs = sdk.Batch()
+    for batch in sdk.util.chunk(prepared_jobs):
+        jobs += HYP3.submit_prepared_jobs(batch)
+
+    return jobs
+
+
+def process_scene(
+    scene: str,
+    max_pair_separation: timedelta = timedelta(days=MAX_PAIR_SEPERATION),
```
Comment (jtherrmann) [inline landsat/src/main.py:165]:
```suggestion
    max_pair_separation: timedelta = timedelta(days=MAX_PAIR_SEPARATION),
```

File: landsat/src/main.py [hunk 4]
```diff
@@ -20,4 +211,35 @@ def lambda_handler(event: dict, context: object) -> None:
...
+
+
+def main() -> None:
+    """Command Line wrapper around `process_scene`."""
+    parser = argparse.ArgumentParser(description=__doc__, formatter_class=argparse.ArgumentDefaultsHelpFormatter)
+    parser.add_argument('reference', help='Reference Landsat scene name to build pairs for')
+    parser.add_argument(
+        '--max-pair-separation',
+        type=int,
+        default=MAX_PAIR_SEPERATION,
```
Comment (jtherrmann) [inline landsat/src/main.py:224]:
```suggestion
        default=MAX_PAIR_SEPARATION,
```

**Review by jtherrmann — DISMISSED** [review]

File: landsat/src/main.py [hunk 5]
```diff
@@ -1,11 +1,199 @@
...
+
+
+def _landsat_tile(scene: str) -> str:
+    return scene.split('_')[2]
+
+
+def _search_date(date_string: str) -> datetime:
+    default_date = datetime(2020, 1, 1, 0, 0, 0, 0, timezone.utc)
+    dt = date_parser(date_string, default=default_date)
+    return dt.astimezone(tz=timezone.utc)
```
Comment (jtherrmann) [inline landsat/src/main.py:58]: What's the reason for having a default date?

File: landsat/src/main.py [hunk 6]
```diff
@@ -1,11 +1,199 @@
...
+        collections=[LANDSAT_COLLECTION],
+        query=[
+            f'landsat:wrs_path={tile[0:3]}',
+            f'landsat:wrs_row={tile[3:]}',
+            f'eo:cloud_cover<{max_cloud_cover}',
+            'view:off_nadir=0',
+        ],
+        datetime=[acquisition_time - max_pair_separation, acquisition_time],
+    )
+    items = [ii for iis in results.pages() for ii in iis]
```
Comment (jtherrmann) [inline landsat/src/main.py:101]:
```suggestion
    items = [item for items in results.pages() for item in items]
```

File: landsat/src/main.py [hunk 7]
```diff
@@ -20,4 +208,35 @@ def lambda_handler(event: dict, context: object) -> None:
     for record in event['Records']:
         body = json.loads(record['body'])
         message = json.loads(body['Message'])
-        process_scene(message['landsat_product_id'])
+        _ = process_scene(message['landsat_product_id'])
```
Comment (jtherrmann) [inline landsat/src/main.py:211]: Why return anything from `process_scene` if we don't use the return value?

File: requirements-landsat.txt
Comment (jtherrmann) [inline requirements-landsat.txt:1]: Should we specify exact versions since we're using dependabot?

File: requirements-all.txt
Comment (jtherrmann) [inline requirements-all.txt:1]: Should we specify exact versions since we're using dependabot?

File: landsat/src/main.py (hunk 5 above)
Comment (jhkennedy) [reply landsat/src/main.py:58]:
```
    :param default:
        The default datetime object, if this is a datetime object and not
        ``None``, elements specified in ``timestr`` replace elements in the
        default object.
```

This ensures that partial dates are in UTC and rounded down. By default, `dateutil` uses `date.today()` to fill in missing days or months, and also doesn't include a timezone in the parsed datetime.

```python
>>> from datetime import datetime, timezone
>>> from dateutil.parser import parse as date_parser
>>> 
>>> date_parser('2024')
datetime.datetime(2024, 3, 1, 0, 0)
>>> 
>>> date_parser('2024', default=datetime(2020, 1, 1, 0, 0, 0, 0, timezone.utc))
datetime.datetime(2024, 1, 1, 0, 0, tzinfo=datetime.timezone.utc)
>>> 
>>> date_parser('20240405')
datetime.datetime(2024, 4, 5, 0, 0)
>>> 
>>> date_parser('20240405', default=datetime(2020, 1, 1, 0, 0, 0, 0, timezone.utc))
datetime.datetime(2024, 4, 5, 0, 0, tzinfo=datetime.timezone.utc)
```

File: landsat/src/main.py (hunk 7 above)
Comment (jhkennedy) [reply landsat/src/main.py:211]: Mostly so I can use the tooling in the interpreter or other scripts when I'm manually processing scenes(e.g., reprocessing failed scenes, filling in missing times, adding new tiles, etc.)

File: requirements-landsat.txt
Comment (jhkennedy) [reply requirements-landsat.txt:1]: yes,  but I'll do that as part of #7 I think.

File: requirements-all.txt
Comment (jhkennedy) [reply requirements-all.txt:1]: yes,  but I'll do that as part of #7 I think.

**Review by forrestfwilliams — APPROVED** [review]

**Pull request merged**